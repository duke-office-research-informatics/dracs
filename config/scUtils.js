import React from "react";
import { ThemeProvider } from "styled-components";
import { mount, shallow } from "enzyme";
import { theme as defaultTheme } from "../lib/dracs.es.js";

export const shallowWithTheme = (children, customTheme) => {
  const theme = customTheme || defaultTheme;
  return shallow(children, { context: theme });
};

/* eslint-disable */
export function mountWithTheme(child) {
  return mount(child, {
    wrappingComponent: ({ children }) => (
      <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
    ),
  });
}

//FOR ACCESSING PORTALS VIA ENZYME -- import portal from dracs
// https://github.com/airbnb/enzyme/issues/1202#issuecomment-351639783
// function mountPortal(el){
//   const wrapper =  mountWithTheme(el);
//   wrapper.setState({ visible: true});
//   const portalInstance = wrapper.find(Portal).instance();
//   return mountWithTheme(portalInstance.props.children);
// }

//FOR ACCESSING EVENT LISTENERS ATTACHED TO DOCUMENT OR DOCUMENT.BODY
//https://github.com/airbnb/enzyme/issues/426#issuecomment-296304015
// const map = {};
// document.body.addEventListener = jest.fn((event, cb) => {
//   map[event] = cb;
// });
