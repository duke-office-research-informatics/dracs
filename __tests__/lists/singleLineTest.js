import React from "react";
import { mountWithTheme } from "../../config/scUtils.js";
import { axe } from "../../config/axe-helper.js";
import {
  List,
  SingleLineListItem,
  theme,
  IconWarning,
} from "../../lib/dracs.es.js";

const title = "test title";
const onLeftIconClick = jest.fn();

describe("Single Line List Item", () => {
  it("matches snapshot w/o a11y violations", async () => {
    const item = mountWithTheme(
      <List>
        <SingleLineListItem />
      </List>
    );
    expect(item).toMatchSnapshot();
    const results = await axe(item.html());
    expect(results).toHaveNoViolations();
  });
  it("sets a border when bottomBorder prop = true", () => {
    const item = mountWithTheme(
      <List>
        <SingleLineListItem bottomBorder />
      </List>
    );
    expect(item).toMatchSnapshot();
    expect(item.find("single_line__ItemWrap")).toHaveStyleRule(
      "border-bottom",
      `solid 1px ${theme.colors.border}`
    );
  });
  it("passes a child to the correct dom node in the list item", () => {
    const item = mountWithTheme(
      <List>
        <SingleLineListItem>
          <IconWarning />
        </SingleLineListItem>
      </List>
    );
    expect(item).toMatchSnapshot();
    expect(item.find(IconWarning)).toBeDefined();
    expect(item.find(IconWarning).length).toBe(1);
    expect(
      item.find("single_line__ItemRightContainer").find(IconWarning)
    ).toBeDefined();
  });
  it("sets the title on the correct dom node via the title prop", () => {
    const item = mountWithTheme(
      <List>
        <SingleLineListItem title={title} />
      </List>
    );
    expect(item).toMatchSnapshot();
    expect(item.find("single_line__TitleWrap").text()).toBe(title);
  });
  it("sets the proper styles when clickable prop = true", async () => {
    const item = mountWithTheme(
      <List>
        <SingleLineListItem title={title} clickable={true} />
      </List>
    );
    expect(item).toMatchSnapshot();
    const results = await axe(item.html());
    expect(results).toHaveNoViolations();
    expect(item.find("single_line__ItemWrap")).toHaveStyleRule(
      "cursor",
      "pointer"
    );
    //work on testing nested hovers
    // console.log(item.find("single_line__ItemWrap").instance());
    // expect(item.find("single_line__ItemWrap")).toHaveStyleRule(
    //   "color",
    //   theme.colors.actionHover,
    //   {
    //     modifier: ":hover .c4",
    //   }
    // );
    // expect(item.find("single_line__TitleWrap")).toHaveStyleRule(
    //   "color",
    //   theme.colors.actionHover,
    //   {
    //     modifier: ".c1:hover .c4",
    //   }
    // );
  });
  it("sets the proper styles when underlineTitleOnHover = true", () => {
    const item = mountWithTheme(
      <List>
        <SingleLineListItem title={title} underlineTitleOnHover />
      </List>
    );
    expect(item).toMatchSnapshot();
    //need to find good way to test hovers on nested component declaration
  });
  it("sets a left icon in the proper dom node when passed as leftIcon prop", () => {
    const item = mountWithTheme(
      <List>
        <SingleLineListItem title={title} leftIcon={<IconWarning />} />
      </List>
    );
    expect(item).toMatchSnapshot();
    expect(item.find(IconWarning)).toBeDefined();
    expect(item.find(IconWarning).length).toBe(1);
    expect(
      item.find("single_line__LeftItemWrap").find(IconWarning)
    ).toBeDefined();
  });
  it("sets a click handler on the left icon via onLeftIconClick prop", async () => {
    const item = mountWithTheme(
      <List>
        <SingleLineListItem
          title={title}
          leftIcon={<IconWarning />}
          onLeftIconClick={onLeftIconClick}
        />
      </List>
    );
    expect(item).toMatchSnapshot();
    const results = await axe(item.html());
    expect(results).toHaveNoViolations();
    const leftIconWrapper = item.find("single_line__LeftItemWrap");
    expect(leftIconWrapper.props().tabIndex).toBe("0");
    leftIconWrapper.simulate("click");
    expect(onLeftIconClick).toHaveBeenCalled();
  });
  it("sets style on the proper dom node when style obj passed via wrapperStyle prop", () => {
    const item = mountWithTheme(
      <List>
        <SingleLineListItem wrapperStyle={{ margin: "8px" }} />
      </List>
    );
    expect(item).toMatchSnapshot();
    expect(item.find("single_line__ItemWrap").props().style).toHaveProperty(
      "margin",
      "8px"
    );
  });
});
