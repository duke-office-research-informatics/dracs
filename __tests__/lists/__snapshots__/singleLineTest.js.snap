// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single Line List Item matches snapshot w/o a11y violations 1`] = `
.c0 {
  padding: 0;
  margin: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 4px 0;
  list-style-type: none;
  width: 100%;
}

.c1:focus {
  outline: dotted 1px rgb(59,153,252);
  outline-offset: -1px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<ul
  className="c0"
>
  <li
    className="c1"
    onKeyUp={[Function]}
    tabIndex={null}
  >
    <div
      className="c2"
    />
    <div
      className=""
    />
  </li>
</ul>
`;

exports[`Single Line List Item passes a child to the correct dom node in the list item 1`] = `
.c3 {
  fill: #ffd960;
  vertical-align: top;
}

.c0 {
  padding: 0;
  margin: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 4px 0;
  list-style-type: none;
  width: 100%;
}

.c1:focus {
  outline: dotted 1px rgb(59,153,252);
  outline-offset: -1px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<ul
  className="c0"
>
  <li
    className="c1"
    onKeyUp={[Function]}
    tabIndex={null}
  >
    <div
      className="c2"
    />
    <div
      className=""
    >
      <svg
        aria-label="warning"
        className="c3"
        color="#ffd960"
        height={24}
        role="img"
        viewBox="0 0 24 24"
        width={24}
      >
        <title>
          warning
        </title>
        <desc>
          triangle with exclamation mark in center
        </desc>
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
        />
      </svg>
    </div>
  </li>
</ul>
`;

exports[`Single Line List Item sets a border when bottomBorder prop = true 1`] = `
.c0 {
  padding: 0;
  margin: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 4px 0;
  list-style-type: none;
  width: 100%;
  border-bottom: solid 1px #e8e5e2;
}

.c1:focus {
  outline: dotted 1px rgb(59,153,252);
  outline-offset: -1px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<ul
  className="c0"
>
  <li
    className="c1"
    onKeyUp={[Function]}
    tabIndex={null}
  >
    <div
      className="c2"
    />
    <div
      className=""
    />
  </li>
</ul>
`;

exports[`Single Line List Item sets a click handler on the left icon via onLeftIconClick prop 1`] = `
.c5 {
  fill: #ffd960;
  vertical-align: top;
}

.c0 {
  padding: 0;
  margin: 0;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 4px 0;
  list-style-type: none;
  width: 100%;
}

.c2:focus {
  outline: dotted 1px rgb(59,153,252);
  outline-offset: -1px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c4 {
  padding: 0 16px 0 8px;
}

.c6 {
  padding: 8px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c1:hover .dZrptj {
  color: #53a4dc;
}

.c1:focus .dZrptj {
  color: #53a4dc;
}

.c1:hover .dbgMsQ {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1:focus .dbgMsQ {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<ul
  className="c0"
>
  <li
    className="c1 c2"
    onKeyUp={[Function]}
    tabIndex={null}
  >
    <div
      className="c3"
    >
      <div
        className="c4"
        onClick={[Function]}
        onKeyUp={[Function]}
        tabIndex="0"
      >
        <svg
          aria-label="warning"
          className="c5"
          color="#ffd960"
          height={24}
          role="img"
          viewBox="0 0 24 24"
          width={24}
        >
          <title>
            warning
          </title>
          <desc>
            triangle with exclamation mark in center
          </desc>
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
          />
        </svg>
      </div>
      <div
        className="c6"
      >
        test title
      </div>
    </div>
    <div
      className=""
    />
  </li>
</ul>
`;

exports[`Single Line List Item sets a left icon in the proper dom node when passed as leftIcon prop 1`] = `
.c5 {
  fill: #ffd960;
  vertical-align: top;
}

.c0 {
  padding: 0;
  margin: 0;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 4px 0;
  list-style-type: none;
  width: 100%;
}

.c2:focus {
  outline: dotted 1px rgb(59,153,252);
  outline-offset: -1px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c4 {
  padding: 0 16px 0 8px;
}

.c6 {
  padding: 8px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c1:hover .dZrptj {
  color: #53a4dc;
}

.c1:focus .dZrptj {
  color: #53a4dc;
}

.c1:hover .dbgMsQ {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1:focus .dbgMsQ {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<ul
  className="c0"
>
  <li
    className="c1 c2"
    onKeyUp={[Function]}
    tabIndex={null}
  >
    <div
      className="c3"
    >
      <div
        className="c4"
        onKeyUp={[Function]}
        tabIndex={null}
      >
        <svg
          aria-label="warning"
          className="c5"
          color="#ffd960"
          height={24}
          role="img"
          viewBox="0 0 24 24"
          width={24}
        >
          <title>
            warning
          </title>
          <desc>
            triangle with exclamation mark in center
          </desc>
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
          />
        </svg>
      </div>
      <div
        className="c6"
      >
        test title
      </div>
    </div>
    <div
      className=""
    />
  </li>
</ul>
`;

exports[`Single Line List Item sets style on the proper dom node when style obj passed via wrapperStyle prop 1`] = `
.c0 {
  padding: 0;
  margin: 0;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 4px 0;
  list-style-type: none;
  width: 100%;
}

.c2:focus {
  outline: dotted 1px rgb(59,153,252);
  outline-offset: -1px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c1:hover .dZrptj {
  color: #53a4dc;
}

.c1:focus .dZrptj {
  color: #53a4dc;
}

.c1:hover .dbgMsQ {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1:focus .dbgMsQ {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<ul
  className="c0"
>
  <li
    className="c1 c2"
    onKeyUp={[Function]}
    style={
      Object {
        "margin": "8px",
      }
    }
    tabIndex={null}
  >
    <div
      className="c3"
    />
    <div
      className=""
    />
  </li>
</ul>
`;

exports[`Single Line List Item sets the proper styles when clickable prop = true 1`] = `
.c0 {
  padding: 0;
  margin: 0;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 4px 0;
  list-style-type: none;
  cursor: pointer;
  width: 100%;
}

.c2:focus {
  outline: dotted 1px rgb(59,153,252);
  outline-offset: -1px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c4 {
  color: #0680cd;
  padding: 8px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c1:hover .c4 {
  color: #53a4dc;
}

.c1:focus .c4 {
  color: #53a4dc;
}

<ul
  className="c0"
>
  <li
    className="c1 c2"
    onKeyUp={[Function]}
    tabIndex={null}
  >
    <div
      className="c3"
    >
      <div
        className="c4"
      >
        test title
      </div>
    </div>
    <div
      className=""
    />
  </li>
</ul>
`;

exports[`Single Line List Item sets the proper styles when underlineTitleOnHover = true 1`] = `
.c0 {
  padding: 0;
  margin: 0;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 4px 0;
  list-style-type: none;
  width: 100%;
}

.c2:focus {
  outline: dotted 1px rgb(59,153,252);
  outline-offset: -1px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c1:hover .dZrptj {
  color: #53a4dc;
}

.c1:focus .dZrptj {
  color: #53a4dc;
}

.c4 {
  padding: 8px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c1:hover .c4 {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1:focus .c4 {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<ul
  className="c0"
>
  <li
    className="c1 c2"
    onKeyUp={[Function]}
    tabIndex={null}
  >
    <div
      className="c3"
    >
      <div
        className="c4"
      >
        test title
      </div>
    </div>
    <div
      className=""
    />
  </li>
</ul>
`;

exports[`Single Line List Item sets the title on the correct dom node via the title prop 1`] = `
.c0 {
  padding: 0;
  margin: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 4px 0;
  list-style-type: none;
  width: 100%;
}

.c1:focus {
  outline: dotted 1px rgb(59,153,252);
  outline-offset: -1px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c3 {
  padding: 8px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<ul
  className="c0"
>
  <li
    className="c1"
    onKeyUp={[Function]}
    tabIndex={null}
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        test title
      </div>
    </div>
    <div
      className=""
    />
  </li>
</ul>
`;
